src/chart_process/bms_processor.rs

```diff
@@ -612,6 +696,136 @@ impl BmsProcessor {
         )
     }
 
+    /// Precompute absolute activate_time for all events based on BPM segmentation and Stops.
+    fn precompute_activate_times(&mut self) {
```

Comment:
I think that it is hard to guarantee that `all_events` is holding the invariants. Also the precomputation process looks too complex. How about to split it into another private struct?